accepted/0000-handling-peer-conflicts.md

```diff
@@ -0,0 +1,337 @@
+failure or a warning.
+
+## Motivation
+
+With the move in npm v7 to [install
+`peerDependencies`](./0025-install-peer-deps.md), npm now has the capacity
+to enter "dependency hell", where a package cannot be installed correctly
+because its `peerDependencies` conflict with other packages in the tree.
+
+For example, given the following dependency graph:
+
+```
+project -> (x@1, y@2)
+x@1 -> PEER(y@1)
+y@2 -> PEER(x@2)
+```
+
+The project cannot correctly install both `x@1` and `y@2` at the same point
+in the dependency tree, because the peer dependency from `x@1` will always
+conflict with the dependency on `y@2`.  And if it does, then the peer
+dependency from `y@2` on `x@2` will conflict with the dependency on `x@1`.
+
+If `project` is published and used as a dependency, there is no way for it
+to get correct dependencies with any arrangement of nesting.  If `project`
+resolves `x` to `x@1`, and `y` is installed at or above `x` in the tree,
+then `x` will resolve to the same version of `y` that `project` does.
+
+Since `peerDependencies` were _not_ installed by default for so long,
+the inclusion of required `peerDependencies` fell to the module with the
+dependency.  Thus, this sort of conflict is widespread within the npm
+package ecosystem.
+
+While the most effective way to get these issues discovered and fixed would
+likely be to fail the install in these situations, it also would make it
+impossible for many users to upgrade to npm v7, which is not ideal.
+
+In many cases, the "expected" resolution can be determined by treating
+`peerDependencies` as lower priority than direct dependencies of the
+package depending on the peer dependency set.
+
+For example, in the conflict described above, we could reasonably infer
+that, at least for the purposes of `project`, using `x@1` and `y@2`
+together is the intent, in spite of their stated conflict, because the
+package bringing in these dependencies has an explictly stated dependency
+on each of them.
+
+In other situations, the expected result is more difficult to determine.
+For example:
+
+```
+project -> (x@1, y@2)
+x@1 -> PEER(z@1)
+y@2 -> PEER(z@2)
+```
+
+In this situation, there is no explicit dependency on `z`, and thus no way
+to know the intent of the author of `project`.
+
+Also, a user encountering a conflict caused by a dependency is in a much
+different position than a user encountering a conflict in their own
+project.  Failing a build due to meta-dependencies is annoying and
+frustrating; failing a build due to incorrect direct dependencies is
+helpful and actionable.
+
+Thus, it is sensible to have different behavior for conflicts caused by
+dependencies than for those caused by the root project or one of its
+workspaces.
+
+## Detailed Explanation
+
+When the `--legacy-peer-deps` flag is set, ignore `peerDependencies`
+entirely.  This makes npm v7 behave effectively the same as npm v3 through
+v6.  The `--legacy-peer-deps` flag is _never_ set for `npm ls`, so that npm
```

Comment:
so `npm ls --legacy-peer-deps` will have the flag as a noop, or will fail as an invalid flag?

Reply:
`ls` passed `legacyPeerDeps: false` in the arborist options regardless of what the config says.